lib/cn/cn_tree.c

```diff
@@ -1356,71 +1359,103 @@ cn_comp_update_kvcompact(struct cn_compaction_work *work, struct kvset *new_kvse
+{
+    atomic_long *node_sgen = &w->cw_node->tn_sgen;
+    struct kvdb_health *hp = w->cw_tree->ct_kvdb_health;
+    bool ready;
+
+    mutex_lock(&tn->tn_spill_mtx);
+
+    while (w->cw_sgen != atomic_read(node_sgen) + 1 &&
+           !atomic_read(w->cw_cancel_request) &&
+           !kvdb_health_check(hp, KVDB_HEALTH_FLAG_ALL)) {
+
+        long timeout_ms = 100;
+        cv_timedwait(&tn->tn_spill_cv, &tn->tn_spill_mtx, timeout_ms, "spillwait");
+    }
+
+    mutex_unlock(&tn->tn_spill_mtx);
+
+    ready = !atomic_read(w->cw_cancel_request) && !kvdb_health_check(hp, KVDB_HEALTH_FLAG_ALL);
+    return ready;
+}
+
 static void
-cn_comp_update_spill(struct cn_compaction_work *work, struct kvset **kvsets)
+cn_node_spill_broadcast(struct cn_tree_node *tn)
+{
+    mutex_lock(&tn->tn_spill_mtx);
+    atomic_inc(&tn->tn_sgen);
+    cv_broadcast(&tn->tn_spill_cv);
+    mutex_unlock(&tn->tn_spill_mtx);
+}
+
+static void
+cn_input_kvsets_delete(struct cn_compaction_work *work)
 {
-    struct cn_tree *         tree = work->cw_tree;
-    struct cn_tree_node *    pnode = work->cw_node;
-    struct kvset_list_entry *le, *tmp;
     struct list_head         retired_kvsets;
-    struct cn_tree_node *    node;
+    struct cn_tree          *tree = work->cw_tree;
+    struct cn_tree_node     *pnode = work->cw_node;
+    struct kvset_list_entry *le, *tmp;
+    struct cndb_txn         *tx;
 
-    if (ev(work->cw_err))
-        return;
+    assert(!work->cw_err);
 
     INIT_LIST_HEAD(&retired_kvsets);
 
-    rmlock_wlock(&tree->ct_lock);
-    {
-        for (uint i = 0; i < work->cw_outc; i++) {
-            if (kvsets[i]) {
-                node = work->cw_output_nodev[i];
-                assert(node);
-                kvset_list_add(kvsets[i], &node->tn_kvset_list);
-                node->tn_cgen++;
-            }
-        }
-
-        /* Advance the change generation on the spill source node
-         * to ensure it is reevaluated by csched/sp3_dirty_node().
-         */
-        pnode->tn_cgen++;
+    work->cw_err = cndb_record_txstart(work->cw_tree->cndb, 0, CNDB_INVAL_INGESTID, CNDB_INVAL_HORIZON,
+                                       0, work->cw_kvset_cnt, &tx);
+    if (work->cw_err)
+        goto done;
 
-        /* Move old kvsets from parent node to retired list.
-         * Asserts:
-         * - Each input kvset just spilled must still be on pnode's kvset list.
-         * - The dgen of the oldest input kvset must match work struct dgen_lo
-         *   (i.e., concurrent spills from a node must be committed in order).
-         */
-        for (uint i = 0; i < work->cw_kvset_cnt; i++) {
-            assert(!list_empty(&pnode->tn_kvset_list));
-            le = list_last_entry(&pnode->tn_kvset_list, struct kvset_list_entry, le_link);
-            assert(i > 0 || work->cw_dgen_lo == kvset_get_dgen(le->le_kvset));
-            list_del(&le->le_link);
-            list_add(&le->le_link, &retired_kvsets);
-        }
+    le = work->cw_mark;
```

Comment:
Maybe useful to assert(le == list_last_entry(&pnode->tn_kvset_list, struct kvset_list_entry, le_link))? This validates that the input kvsets are deleted in order.